tt_metal/impl/dispatch/command_queue_interface.hpp

```diff
@@ -157,32 +163,66 @@ inline uint32_t get_cq_completion_wr_ptr(chip_id_t chip_id, uint8_t cq_id, uint3
 // Ideally would work by cachelines, but the min size is less than that
 // TODO: Revisit this w/ regard to possibly eliminating min sizes and orphan writes at the end
 // TODO: ditto alignment isues
-// TODO: move _sfence out
+template <bool blocking = false>
 static inline void memcpy_to_device(void *__restrict dst, const void * __restrict src, size_t n)
 {
-    if (n < CQ_PREFETCH_CMD_BARE_MIN_SIZE) {
-        memcpy(dst, src, n);
-        return;
-    }
+    TT_ASSERT((uintptr_t)dst % MEMCPY_ALIGNMENT == 0);
+    TT_ASSERT(n % sizeof(uint32_t) == 0);
 
-    size_t num_lines = n / CQ_PREFETCH_CMD_BARE_MIN_SIZE;
+    static constexpr uint32_t inner_loop = 8;
+    static constexpr uint32_t inner_blk_size = inner_loop * sizeof(__m128i);
 
     uint8_t *src8 = (uint8_t *)src;
     uint8_t *dst8 = (uint8_t *)dst;
-    for (size_t i = 0; i < num_lines; i++) {
-        for (size_t j = 0; j < CQ_PREFETCH_CMD_BARE_MIN_SIZE / sizeof(__m128i); j++) {
-            __m128i blk = _mm_loadu_si128((const __m128i *)src);
-            _mm_stream_si128((__m128i *)dst, blk);
+    if (n < inner_blk_size) {
+        for (size_t i = 0; i < n / sizeof(__m128i); i++) {
+            __m128i blk = _mm_loadu_si128((const __m128i *)src8);
+            _mm_stream_si128((__m128i *)dst8, blk);
             src8 += sizeof(__m128i);
             dst8 += sizeof(__m128i);
         }
-        n -= CQ_PREFETCH_CMD_BARE_MIN_SIZE;
+        n -= n / sizeof(__m128i) * sizeof(__m128i);
+        int32_t *src32 = (int32_t *)src8;
+        int32_t *dst32 = (int32_t *)dst8;
+        for (size_t i = 0; i < n / sizeof(int32_t); i++) {
+            _mm_stream_si32(dst32 + i, src32[i]);
+        }
+        if constexpr(blocking) {
+            tt_driver_atomics::sfence();
```

Comment:
Currently when running there is only a sfence before writing the fetchq, I've removed the other existing fences and we call this with the template blocking=false. I left this here for testing since it's constexpr, but I can remove the templating entirely if we want.

Reply:
ah, got it. may rename "blocking" to "debug_sync" or some such